### PR TITLE
support for getting time from JDBC-enabled database

### LINKED ISSUE #3383: Implement Orcale，PostgreSQL and SQLServer time service
## Feature Request

**For English only**, other languages will not accept.

Please pay attention on issues you submitted, because we maybe need more details. 
If no response **more than 7 days** and we cannot make decision by current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Describe the feature you would like.
- Implement Orcale，PostgreSQL and SQLServer time service
- Automatically select the `TimeService` if import `database-time-service` module
- Configure the connection pool by user

### PR DESCRIPTION
**Pull request by Pramy**
Fixes #3383 

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

File: sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/DataBaseTimeService.java [hunk 1]
```diff
@@ -34,45 +34,48 @@
...
             Class dataSourceClass = Class.forName(dataSourceType);
             DataSource dataSource = (DataSource) dataSourceClass.newInstance();
             for (String each : properties.stringPropertyNames()) {
                 PropertyDescriptor propertyDescriptor = new PropertyDescriptor(each, dataSourceClass);
                 Method writeMethod = propertyDescriptor.getWriteMethod();
                 writeMethod.invoke(dataSource, properties.getProperty(each));
             }
-            MySQLTimeService.dataSource = dataSource;
+            DataBaseTimeService.dataSource = dataSource;
         } catch (final NullPointerException | IOException | ClassNotFoundException | InstantiationException | IllegalAccessException | InvocationTargetException | IntrospectionException e) {
```
Comment (terrymanu) [inline sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/DataBaseTimeService.java]: Why catch NullPointerException?

**Comment by Pramy** [comment]
> Who will set the value of the SQL dialect to fetch time?

I think user should set sql in the time-service.properties,such as `sql=select now()`

**Comment by terrymanu** [comment]
> > Who will set the value of the SQL dialect to fetch time?
> 
> I think user should set sql in the time-service.properties,such as `sql=select now()`

Can we add some default SQLs for known databases?

**Comment by Pramy** [comment]
> > > Who will set the value of the SQL dialect to fetch time?
> > 
> > 
> > I think user should set sql in the time-service.properties,such as `sql=select now()`
> 
> Can we add some default SQLs for known databases?
> And does the `time-service.properties` support for multiple database dialects? eg: `key(database name)` = `value(select now())`?

hi, I have implemented  MySQL, Oracle, SQLServer and PostgreSQL database time service, otherwise, I link mariab successfully. As long as we have a JDBC Driver, we can link multiple databases.

File: sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/DataBaseTimeServiceDelegate.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.shardingsphere.core.route.spi.TimeService;
+
+import java.util.Date;
+
+/**
+ * Delegate database time service.
+ *
+ * @author chenchuangliu
+ */
+public final class DataBaseTimeServiceDelegate implements TimeService {
```
Comment (terrymanu) [inline sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/DataBaseTimeServiceDelegate.java]: `Database` is one word, so please rename `DataBaseTimeServiceDelegate` to `DatabaseTimeServiceDelegate`

File: sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/impl/BaseDataBaseTimeService.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.Date;
+
+/**
+ * Base database time service.
+ *
+ * @author chenchuangliu
+ */
+@RequiredArgsConstructor
+@Getter(AccessLevel.PACKAGE)
+public abstract class BaseDataBaseTimeService implements TimeService {
```
Comment (terrymanu) [inline sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/impl/BaseDataBaseTimeService.java]: `Database` is one word, so please rename `BaseDataBaseTimeService` to `BaseDataBaseTimeService`

File: sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/impl/BaseDataBaseTimeService.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    private final DataSource dataSource;
+
+    private final String sql;
+
+    @Override
+    public final Date getTime() {
+        if (null != dataSource) {
```
Comment (terrymanu) [inline sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/impl/BaseDataBaseTimeService.java]: When dataSource will be null?

File: sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/impl/BaseDataBaseTimeService.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            try (Connection connection = dataSource.getConnection();
+                 PreparedStatement preparedStatement = connection.prepareStatement(null == sql ? getTimeSql() : sql);
+                 ResultSet resultSet = preparedStatement.executeQuery()) {
```
Comment (terrymanu) [inline sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/impl/BaseDataBaseTimeService.java]: Need another try with resource for `ResultSet`

File: sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/impl/BaseDataBaseTimeService.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                resultSet.next();
+                return (Date) resultSet.getObject(1);
+            } catch (final SQLException ignore) {
+            }
+        }
+        return new Date();
+    }
+
+    /**
+     * Get sql for getting time.
```
Comment (terrymanu) [inline sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/impl/BaseDataBaseTimeService.java]: `SQL` should always be upper case in java doc

File: sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/impl/BaseDataBaseTimeService.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+     * @return sql
+     */
+    public abstract String getTimeSql();
```
Comment (terrymanu) [inline sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/impl/BaseDataBaseTimeService.java]: Rename `getTimeSql` to `getTimeSQL` is better

**Comment by terrymanu** [comment]
Perfect pull request

**Pull request merged**